src/clientrpc.cpp

```diff
@@ -0,0 +1,166 @@
+                                     { "name", chanInfo.strName },
+                                     { "skillLevel", FormatSkillLevel ( chanInfo.eSkillLevel ) },
+                                     { "countryId", chanInfo.eCountry },
+                                     { "city", chanInfo.strCity },
+                                     { "instrumentId", chanInfo.iInstrument } };
+            arrChanInfo.append ( objChanInfo );
+        }
+        pRpcServer->BroadcastNotification ( "jamulusclient/clientListReceived", QJsonObject{ { "clients", arrChanInfo } } );
+        arrStoredChanInfo = arrChanInfo;
+    } );
+
+    connect ( pClient, &CClient::CLChannelLevelListReceived, [=] ( CHostAddress /* unused */, CVector<uint16_t> vecLevelList ) {
+        QJsonArray arrLevelList;
+        for ( const auto& level : vecLevelList )
+        {
+            arrLevelList.append ( level );
+        }
+        pRpcServer->BroadcastNotification ( "jamulusclient/channelLevelListReceived", QJsonObject{ { "channelLevelList", arrLevelList } } );
+    } );
+
+    connect ( pClient, &CClient::Disconnected, [=]() { pRpcServer->BroadcastNotification ( "jamulusclient/disconnected", QJsonObject{} ); } );
+
+    pRpcServer->HandleMethod ( "jamulusclient/getClientInfo", [=] ( const QJsonObject& params, QJsonObject& response ) {
+        QJsonObject result{ { "connected", pClient->IsConnected() } };
+        response["result"] = result;
+    } );
+
+    pRpcServer->HandleMethod ( "jamulusclient/getChannelInfo", [=] ( const QJsonObject& params, QJsonObject& response ) {
+        QJsonObject result{ { "name", pClient->ChannelInfo.strName }, { "skillLevel", FormatSkillLevel ( pClient->ChannelInfo.eSkillLevel ) } };
+        response["result"] = result;
+    } );
+
+    pRpcServer->HandleMethod ( "jamulusclient/getClientList", [=] ( const QJsonObject& params, QJsonObject& response ) {
+        QJsonObject result{ { "clients", arrStoredChanInfo } };
+        response["result"] = result;
+    } );
+
+    pRpcServer->HandleMethod ( "jamulusclient/setName", [=] ( const QJsonObject& params, QJsonObject& response ) {
+        auto jsonName = params["name"];
+        if ( !jsonName.isString() )
+        {
+            response["error"] = CreateJsonRpcError ( -32602, "Invalid params" );
+            return;
+        }
+
+        pClient->ChannelInfo.strName = jsonName.toString().left ( MAX_LEN_FADER_TAG );
+        pClient->SetRemoteInfo();
+
+        response["result"] = "ok";
+    } );
+
+    pRpcServer->HandleMethod ( "jamulusclient/setSkillLevel", [=] ( const QJsonObject& params, QJsonObject& response ) {
+        auto jsonSkillLevel = params["skillLevel"];
+        if ( jsonSkillLevel.isNull() )
+        {
+            pClient->ChannelInfo.eSkillLevel = SL_NOT_SET;
+            pClient->SetRemoteInfo();
+            return;
+        }
+
+        if ( !jsonSkillLevel.isString() )
+        {
+            response["error"] = CreateJsonRpcError ( -32602, "Invalid params" );
```

Comment:
See, as a software engineer who works in support, I don't like seeing two error responses that look identical but mean different things.  If someone shows me the error id and/or text, I want to be able to use that as a unique identifier to find what went wrong.  Here, you've used `-32602` to mean `"Invalid params"` wherever it happens.  I'd much rather see separate errors for each case.

The same applies to the other error cases sharing an id and message.

Reply:
Done — error messages improved

1d169c2d9b243b511a77e9fb50cc5a93d4624718